hazelcast/src/test/java/com/hazelcast/internal/partition/impl/MigrationCommitServiceTest.java

```diff
@@ -102,12 +102,18 @@ public void run() {
                 for (int partitionId = 0; partitionId < PARTITION_COUNT; partitionId++) {
                     InternalPartitionService partitionService = getPartitionService(instances[0]);
                     InternalPartition partition = partitionService.getPartition(partitionId);
-                    for (int i = 0; i <= BACKUP_COUNT; i++) {
+
+                    // given that NODE_COUNT < BACKUP_COUNT, there can be at most
+                    // NODE_COUNT replicas overall.
+                    final int replicasCount = Math.min(BACKUP_COUNT + 1, NODE_COUNT);
+                    for (int i = 0; i < replicasCount; i++) {
+                        // replica assignment should complete, so that when we later
+                        // reference a replica, we do not get NPE.
```

Comment:
sorry, maybe i'm missing something, which line was throwing NPE and how does this change fix it?

Reply:
Here, https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/test/java/com/hazelcast/internal/partition/impl/MigrationCommitServiceTest.java#L361 it throws `NullPointerException` because `oldReplicaOwner` is null which is declared couple of lines above as `final PartitionReplica oldReplicaOwner = partition.getReplica(replicaIndexToClear);`

Reply:
Thanks i see now. 
In the issue, the failing test with NPE is `testPartitionOwnerShiftDownCommit`.(i've ran it on master and hasn't failed, may be a sporadical failure?)
```
@Test
    public void testPartitionOwnerShiftDownCommit() throws Exception {
        int oldReplicaIndex = 0;
        int replicaIndexToClear = NODE_COUNT - 1;
        int newReplicaIndex = NODE_COUNT - 1;
        testSuccessfulShiftDownMigration(PARTITION_ID_TO_MIGRATE, replicaIndexToClear, oldReplicaIndex, newReplicaIndex);
    }
```

As i see from the test above, with or without the change in this pr, replica indexes are always smaller than `NODE_COUNT`. This is why i couldn't correlate the fix and the issue and couldn't see how your change affects this test.

Reply:
Thanks :) Yeah it is kinda flaky and almost never fails, but it did few times on my local machine. Regrading the fix, it is in the following lines

    // currently
    if (replicaAddress != null) {
           TestMigrationAwareService service = getService(replicaAddress);
           assertNotNull(service.get(partitionId));
    }

we remove the if condition and make sure replicaAddress is never null, i.e., every replica should have an owner. And the reasoning behind decreasing the upper bound to `NODE_COUNT` is because, for example say our `NODE_COUNT = 4`, but `BACKUP_COUNT = 7`, then we cannot have any node own `replicaIndex >= 4`, but for each `replicaIndex < 4` we want it to have an owner so that when we later try to remove old replicaOwner for a partition we don't face scenario where partition does not have owner yet, or its backup does not have owner.
Please let me know if I am missing something or the bigger picture.

Reply:
```
// currently
if (replicaAddress != null) {
       TestMigrationAwareService service = getService(replicaAddress);
       assertNotNull(service.get(partitionId));
}
```
Regardless of the removal of if condition, i see nothing forces replicaAddress to a non-null value. If replicaAddress is null previously it will continue to be null in your new code. Can you point me how we force a non-null replicaAddress?

Reply:
Ha yeah, here it is https://github.com/hazelcast/hazelcast/pull/18975/files#diff-03b7f3ce24b05b07290854d107b226458b207dfc10f65ef83365cd7bf6450b01R113

if it fails `assertTrueEventually` that encapsulates whole block will keep retrying

Reply:
Seems to me NPE can happen, for example, in a split brain like scenario. Otherwise, in sunny days, assignment of first replica means others are also assigned. For split brain case, retrying will not help always since there may be no such replica and null can be an expected value. Another thing, even split can occur after `setup` method is run and we can still see NPE. Not a perfect solution i see but as an enhancement we can try to decrease load while test is running by using `smallInstanceConfig`, this can make splits less likely.

Reply:
I see, I just thought making sure replica assignments are complete at least on the setup phase before test runs, it should provide some more safety. Let me look into `smallInstanceConfig`, thanks for the insight and the suggestion.

Reply:
@ahmetmircik hi, I applied your `new Config() -> smallInstanceConfig()` suggestion.